data/keystore/src/test/java/tech/pegasys/artemis/bls/keystore/KeyStoreTest.java

```diff
@@ -0,0 +1,262 @@
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final KeyStore encryptedKeyStore =
+        KeyStore.encrypt(BLS_PRIVATE_KEY, "test", "", kdfParam, cipher);
+    assertTrue(encryptedKeyStore.validatePassword("test"));
+    final Bytes decryptedKey = encryptedKeyStore.decrypt("test");
+    assertEquals(BLS_PRIVATE_KEY, decryptedKey);
+  }
+
+  @ParameterizedTest
+  @EnumSource(CryptoFunction.class)
+  void encryptLargeDataWithRandomSaltAndIv(final CryptoFunction cryptoFunction) {
+    final KdfParam kdfParam =
+        cryptoFunction == SCRYPT
+            ? SCryptParamBuilder.aSCryptParam().build()
+            : Pbkdf2ParamBuilder.aPbkdf2Param().build();
+    final CipherParam cipherParam = CipherParamBuilder.aCipherParam().build();
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final Bytes largeData = Bytes.random(1024);
+    final Crypto crypto =
+        KeyStore.encryptUsingCipherFunction(
+            largeData, PASSWORD, kdfParam, cipher, ChecksumFunction.SHA256);
+    assertEquals(32, crypto.getChecksum().getMessage().size());
+  }
+
+  @ParameterizedTest
+  @EnumSource(CryptoFunction.class)
+  void encryptSmallDataWithRandomSaltAndIv(final CryptoFunction cryptoFunction) {
+    final KdfParam kdfParam =
+        cryptoFunction == SCRYPT
+            ? SCryptParamBuilder.aSCryptParam().build()
+            : Pbkdf2ParamBuilder.aPbkdf2Param().build();
+    final CipherParam cipherParam = CipherParamBuilder.aCipherParam().build();
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final Bytes smallData = Bytes.random(8);
+    final Crypto crypto =
+        KeyStore.encryptUsingCipherFunction(
+            smallData, PASSWORD, kdfParam, cipher, ChecksumFunction.SHA256);
+    assertEquals(32, crypto.getChecksum().getMessage().size());
+  }
+
+  @ParameterizedTest
+  @EnumSource(CryptoFunction.class)
+  void checksumFunctionSha512(final CryptoFunction cryptoFunction) {
+    final KdfParam kdfParam =
+        cryptoFunction == SCRYPT
+            ? SCryptParamBuilder.aSCryptParam().build()
+            : Pbkdf2ParamBuilder.aPbkdf2Param().build();
+    final CipherParam cipherParam = CipherParamBuilder.aCipherParam().build();
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final Bytes smallData = Bytes.random(8);
+    final Crypto crypto =
+        KeyStore.encryptUsingCipherFunction(
+            smallData, PASSWORD, kdfParam, cipher, ChecksumFunction.SHA512);
+    assertEquals(64, crypto.getChecksum().getMessage().size());
```

Comment:
This also seems like too weak an assertion, suggesting we're not really testing the function.